internal/idtokenvalidator/idtokenvalidator.go

```diff
@@ -0,0 +1,202 @@
+package idtokenvalidator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	"github.com/lestrrat-go/jwx/v2/jws"
+	"github.com/lestrrat-go/jwx/v2/jwt"
+)
+
+// ValidationOptions allows validating optional claims that might not always be in the ID token.
+type ValidationOptions struct {
+	MaxAge         time.Duration
+	Nonce          string
+	OrganizationID string
+}
+
+// IDTokenValidator is used to validate ID tokens retrieved from Auth0.
+type IDTokenValidator struct {
+	alg            jwa.SignatureAlgorithm
+	audience       string
+	clientSecret   []byte
+	clockTolerance time.Duration
+	httpClient     *http.Client
+	issuer         string
+	jwks           *jwk.Cache
+	jwksURL        string
+}
+
+// New creates and returns a new IDTokenValidator.
+func New(
+	ctx context.Context,
+	domain string,
+	clientID string,
+	clientSecret string,
+	idTokenSigningAlg string,
+	opts ...Option,
+) (*IDTokenValidator, error) {
+	if i := strings.Index(domain, "//"); i != -1 {
+		domain = domain[i+2:]
+	}
+	alg, err := determineAlg(idTokenSigningAlg)
+	if err != nil {
+		return nil, err
+	}
+
+	i := &IDTokenValidator{
+		clientSecret:   []byte(clientSecret),
+		alg:            alg,
+		clockTolerance: time.Minute,
+		issuer:         "https://" + domain + "/",
+		audience:       clientID,
+	}
+
+	for _, option := range opts {
+		option(i)
+	}
+
+	if alg == jwa.RS256 {
+		i.jwksURL = i.issuer + ".well-known/jwks.json"
+		i.jwks = jwk.NewCache(ctx)
+		registerOpts := []jwk.RegisterOption{}
+		if i.httpClient != nil {
+			registerOpts = append(registerOpts, jwk.WithHTTPClient(i.httpClient))
+		}
+
+		err = i.jwks.Register(i.jwksURL, registerOpts...)
+		if err != nil {
+			return nil, err
+		}
+
+		_, err = i.jwks.Refresh(ctx, i.jwksURL)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return i, nil
+}
+
+// Validate validates the provided ID token against the values provided during the IDTokenValidator creation.
+func (i *IDTokenValidator) Validate(idToken string, optional ValidationOptions) error {
+	validator := jwt.ValidatorFunc(func(_ context.Context, t jwt.Token) jwt.ValidationError {
+		if t.Subject() == "" {
```

Comment:
Has this been superseded by `jwt.WithRequiredClaim("sub")`?

Reply:
Annoyingly not as it still allows an empty string to pass through

Reply:
So, should we check that `iss` is not empty a well?

Reply:
Because we pass `jwt.WithIssuer(i.issuer)` then an empty string will return an error.

I'm kinda uncertain around the split there is between some validation in jwx and the custom validator as it might lead to something being missed, now there's a solid test suite I might take a look before GA at usefulness of owning all the validation (and disabling jwx validation) vs what there is now.